flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/AdaptiveJoinProcessor.java

```diff
@@ -0,0 +1,176 @@
+                        // distribution type is KEEP_INPUT_AS_IS, as it may cause data
+                        // incorrectness.
+                        if (shouldKeepInputAsIs(node.getInputProperties())) {
+                            return;
+                        }
+                        for (int i = 0; i < node.getInputEdges().size(); ++i) {
+                            ExecEdge edge = node.getInputEdges().get(i);
+                            ExecNode<?> newNode = tryReplaceWithAdaptiveJoinNode(edge.getSource());
+                            node.replaceInputEdge(
+                                    i,
+                                    ExecEdge.builder()
+                                            .source(newNode)
+                                            .target(node)
+                                            .shuffle(edge.getShuffle())
+                                            .exchangeMode(edge.getExchangeMode())
+                                            .build());
+                        }
+                    }
+                };
+
+        List<ExecNode<?>> newRootNodes =
+                execGraph.getRootNodes().stream()
+                        .map(
+                                node -> {
+                                    node = tryReplaceWithAdaptiveJoinNode(node);
+                                    node.accept(visitor);
+                                    return node;
+                                })
+                        .collect(Collectors.toList());
+
+        return new ExecNodeGraph(execGraph.getFlinkVersion(), newRootNodes);
+    }
+
+    private ExecNode<?> tryReplaceWithAdaptiveJoinNode(ExecNode<?> node) {
+        // For AdaptiveJoin to be converted, its upstream input must ensure:
+        // 1. Data distribution is by Hash.
+        // 2. No upstream nodes require KEEP_INPUT_AS_IS (achieved by inserting BatchExecExchange).
+        if (!(areAllInputsHashShuffle(node))
+                || shouldKeepUpstreamNodeInputAsIs(node.getInputEdges())) {
+            return node;
+        }
+        ExecNode<?> newNode = node;
+        if (node instanceof AdaptiveJoinExecNode
+                && ((AdaptiveJoinExecNode) node).canBeTransformedToAdaptiveJoin()) {
+            BatchExecAdaptiveJoin adaptiveJoin = ((AdaptiveJoinExecNode) node).toAdaptiveJoinNode();
+            replaceInputEdge(adaptiveJoin, node);
+            newNode = adaptiveJoin;
+        }
+
+        return newNode;
+    }
+
+    private boolean shouldKeepInputAsIs(List<InputProperty> inputProperties) {
+        return inputProperties.stream()
+                .anyMatch(
+                        inputProperty ->
+                                inputProperty.getRequiredDistribution().getType()
+                                        == KEEP_INPUT_AS_IS);
+    }
+
+    private boolean shouldKeepUpstreamNodeInputAsIs(List<ExecEdge> inputEdges) {
```

Comment:
Maybe `shouldKeepUpstreamExchangeInputAsIs` and add comment to explain that if `KEEP_INPUT_AS_IS` constraint exists on an operator, it will always show on its upstream exchange.